validator/nitro_machine.go

```diff
@@ -17,101 +17,105 @@ import (
-	if moduleRoot == (common.Hash{}) {
-		return filepath.Join(c.RootPath, "latest")
+	if rootPath != "" {
+		places = append(places, rootPath)
 	} else {
-		return filepath.Join(c.RootPath, moduleRoot.String())
-	}
-}
+		// Check the project dir: <project>/arbnode/node.go => ../../target/machines
+		_, thisFile, _, ok := runtime.Caller(0)
+		if !ok {
+			panic("failed to find root path")
+		}
+		projectDir := filepath.Dir(filepath.Dir(thisFile))
+		projectPath := filepath.Join(filepath.Join(projectDir, "target"), "machines")
+		places = append(places, projectPath)
+
+		// Check the working directory: ./machines and ./target/machines
+		workDir, err := os.Getwd()
+		if err != nil {
+			return nil, err
+		}
+		workPath1 := filepath.Join(workDir, "machines")
+		workPath2 := filepath.Join(filepath.Join(workDir, "target"), "machines")
+		places = append(places, workPath1)
+		places = append(places, workPath2)
 
-func (c NitroMachineConfig) ReadLatestWasmModuleRoot() (common.Hash, error) {
-	fileToRead := filepath.Join(c.getMachinePath(common.Hash{}), "module-root.txt")
-	fileBytes, err := os.ReadFile(fileToRead)
-	if err != nil {
-		return common.Hash{}, err
+		// Check above the executable: <binary> => ../../machines
+		execfile, err := os.Executable()
+		if err != nil {
+			return nil, err
+		}
+		execPath := filepath.Join(filepath.Dir(filepath.Dir(execfile)), "machines")
+		places = append(places, execPath)
+	}
+
+	for _, place := range places {
+		if _, err := os.Stat(place); err == nil {
+			var latestModuleRoot common.Hash
+			latestModuleRootPath := filepath.Join(place, "latest", "module-root.txt")
+			fileBytes, err := os.ReadFile(latestModuleRootPath)
+			if err == nil {
+				s := strings.TrimSpace(string(fileBytes))
+				latestModuleRoot = common.HexToHash(s)
+			}
+			return &MachineLocator{place, latestModuleRoot}, nil
+		}
 	}
-	s := strings.TrimSpace(string(fileBytes))
-	return common.HexToHash(s), nil
+	return nil, ErrMachineNotFound
 }
 
-type loaderMachineStatus struct {
-	machine    *ArbitratorMachine
-	jitMachine *JitMachine
-	chanSignal chan struct{}
-	jit        bool
-	err        error
+func (l MachineLocator) getMachinePath(moduleRoot common.Hash) string {
+	if moduleRoot == (common.Hash{}) || moduleRoot == l.latest {
+		return filepath.Join(l.rootPath, "latest")
+	} else {
+		return filepath.Join(l.rootPath, moduleRoot.String())
+	}
 }
 
-func (s *loaderMachineStatus) signalReady() {
-	close(s.chanSignal)
+func (l MachineLocator) LatestWasmModuleRoot() common.Hash {
+	return l.latest
 }
 
-func (s *loaderMachineStatus) createZeroStepMachineInternal(config NitroMachineConfig, moduleRoot common.Hash, realModuleRoot common.Hash) {
-	defer s.signalReady()
-	binPath := filepath.Join(config.getMachinePath(moduleRoot), config.WavmBinaryPath)
+func createArbMachineThread(ctx context.Context, locator *MachineLocator, config *ArbitratorMachineConfig, moduleRoot common.Hash, status *machineStatus[arbMachines]) {
```

Comment:
The naming of the function confused me a bit, because I thought it was creating a thread for arbitrum machines, rather than being what we run in the existing thread. I'd recommend just calling it `createArbMachine` or `createArbMachineHandler` or adding a comment if you keep the existing name.

Reply:
agreed